x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.optimizer;
+
+import org.elasticsearch.xpack.esql.plan.logical.Enrich;
+import org.elasticsearch.xpack.esql.plan.logical.Eval;
+import org.elasticsearch.xpack.esql.plan.logical.MvExpand;
+import org.elasticsearch.xpack.esql.plan.logical.RegexExtract;
+import org.elasticsearch.xpack.esql.plan.logical.Row;
+import org.elasticsearch.xpack.esql.plan.logical.local.LocalRelation;
+import org.elasticsearch.xpack.esql.plan.physical.AggregateExec;
+import org.elasticsearch.xpack.esql.plan.physical.EnrichExec;
+import org.elasticsearch.xpack.esql.plan.physical.EsQueryExec;
+import org.elasticsearch.xpack.esql.plan.physical.EsSourceExec;
+import org.elasticsearch.xpack.esql.plan.physical.EsStatsQueryExec;
+import org.elasticsearch.xpack.esql.plan.physical.EvalExec;
+import org.elasticsearch.xpack.esql.plan.physical.ExchangeExec;
+import org.elasticsearch.xpack.esql.plan.physical.ExchangeSourceExec;
+import org.elasticsearch.xpack.esql.plan.physical.FieldExtractExec;
+import org.elasticsearch.xpack.esql.plan.physical.LocalSourceExec;
+import org.elasticsearch.xpack.esql.plan.physical.MvExpandExec;
+import org.elasticsearch.xpack.esql.plan.physical.PhysicalPlan;
+import org.elasticsearch.xpack.esql.plan.physical.RegexExtractExec;
+import org.elasticsearch.xpack.esql.plan.physical.RowExec;
+import org.elasticsearch.xpack.esql.plan.physical.ShowExec;
+import org.elasticsearch.xpack.ql.common.Failure;
+import org.elasticsearch.xpack.ql.expression.AttributeSet;
+import org.elasticsearch.xpack.ql.expression.Expressions;
+import org.elasticsearch.xpack.ql.plan.QueryPlan;
+import org.elasticsearch.xpack.ql.plan.logical.Aggregate;
+import org.elasticsearch.xpack.ql.plan.logical.EsRelation;
+import org.elasticsearch.xpack.ql.plan.logical.LogicalPlan;
+
+import java.util.Collection;
+
+import static org.elasticsearch.xpack.ql.common.Failure.fail;
+
+class OptimizerRules {
+
+    private OptimizerRules() {}
+
+    static class DependencyConsistency<P extends QueryPlan<P>> {
+
+        void checkPlan(P p, Collection<Failure> failures) {
+            AttributeSet refs = references(p);
+            AttributeSet input = p.inputSet();
+            AttributeSet generated = generates(p);
+            AttributeSet missing = refs.subtract(input).subtract(generated);
```

Comment:
At a first look I thought the usage of "generated" here was to account for intermediate values reused during the command calculation, eg. `EVAL a = 12, b = a + 2`

Looking closer I think it's not the case.

Debugging it I noticed that, for some commands, `references(p)` cannot distinguish between "used" and "emitted" values;
Eg. for MV_EXPAND `references(p)` returns both the expanding and the expanded attribute, that's why when you calculate the `missing` you have to remove the "generated".

This mechanism does not seem rock solid (ie. a command could pass this validation also declaring one of its needed inputs as a generated output), but I think it's a good approximation and a significant step forward in terms of plan validation.

Reply:
I agree that it's weird that for `MV_EXPAND x`, the expanded attribute for `x` will show up both in `generated` and in `refs` - maybe that's a bug in `MV_EXPAND`, and its `references` should not return the expanded attribute.

For `EVAL a = 12, b = a + 2` the constant folding takes care of the re-used `a`, but I found that for `from testidx | eval a = x +1, y = least(a, x)`, the re-used `a` shows up correctly in the refs _and_ the generated attributes.